src/Adapter/Carrier/Repository/CarrierRangeRepository.php

```diff
@@ -0,0 +1,249 @@
+            ->andWhere('cd.id_carrier = :carrierId')
+            ->setParameter('carrierId', $carrierId->getValue());
+
+        // Apply shipping method (weight or price) and shop constraint, then execute the query
+        $this->applyRangeTypeForQuery($qb, $carrier);
+        $this->applyShopConstraint($qb, $shopConstraint);
+        $results = $qb->executeQuery()->fetchAllAssociative();
+
+        // Format results with CarrierRange and CarrierRangePrice objets
+        return self::formatRangesFromData($results);
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     * @param array $ranges
+     * @param ShopConstraint $shopConstraint
+     *
+     * @return CarrierId
+     */
+    public function set(CarrierId $carrierId, array $ranges, ShopConstraint $shopConstraint): CarrierId
+    {
+        // Check shop constraint
+        $this->assertShopConstraint($shopConstraint);
+
+        // Get carrier
+        $carrier = $this->carrierRepository->get($carrierId);
+        $rangeTable = $this->getRangeTypeTable($carrier);
+
+        // Use transaction to ensure data consistency
+        $this->connection->beginTransaction();
+
+        // Reset carrier ranges
+        $this->reset($carrierId, $shopConstraint);
+
+        // Foreach range
+        foreach ($ranges as $range) {
+            // Insert range
+            $this->connection->insert(
+                $this->dbPrefix . $rangeTable,
+                [
+                    'id_carrier' => $carrierId->getValue(),
+                    'delimiter1' => $range->getFrom(),
+                    'delimiter2' => $range->getTo(),
+                ]
+            );
+            $rangeId = $this->connection->lastInsertId();
+
+            // Foreach prices in range
+            foreach ($range->getPrices() as $price) {
+                // Insert price
+                $this->connection->insert(
+                    $this->dbPrefix . 'delivery',
+                    [
+                        'id_carrier' => $carrierId->getValue(),
+                        'id_' . $rangeTable => $rangeId,
+                        'id_zone' => $price->getZoneId(),
+                        'price' => $price->getPrice(),
+                    ]
+                );
+            }
+        }
+
+        // Commit transaction
+        $this->connection->commit();
+
+        return $carrierId;
+    }
+
+    /**
+     * Format ranges with CarrierRange and CarrierRangePrice objects from data
+     *
+     * @param IteratorAggregate|array $iterator
+     *
+     * @return CarrierRange[]
+     *
+     * @throws \PrestaShop\PrestaShop\Core\Domain\Zone\Exception\ZoneException
+     */
+    public static function formatRangesFromData(IteratorAggregate|array $iterator): array
+    {
+        $ranges = $prices = [];
+        $tmpLast = ['from' => null, 'to' => null];
+        foreach ($iterator as $range) {
+            if ($tmpLast['from'] !== $range['range_from'] || $tmpLast['to'] !== $range['range_to']) {
```

Comment:
Not sure I understand this part but it will be easier to discuss it live

Reply:
To document more